src/tezos.ts

```diff
@@ -250,12 +334,109 @@ export class TezosHelpers implements Helpers {
+      txBlockDetails.numEndorsements === 32
+    ) {
       res.status = "Confirmed";
     }
 
     return res;
   }
+
+  encodePublicKey(pubKey: string): string {
+    const pref = {
+      ed: prefix["edpk"],
+      p2: prefix["p2pk"],
+      sp: prefix["sppk"],
+    };
+    const curve = pubKey.substring(0, 2);
+    if (Object.prototype.hasOwnProperty.call(pref, curve))
+      return Buffer.from(
+        b58cdecode(pubKey, Object.getOwnPropertyDescriptor(pref, curve)?.value),
+      ).toString("hex");
+    throw new Error("Unsupported Public Key Type");
+  }
+
+  encodeSignature(signature: string): string {
+    const pref = signature.startsWith("sig")
+      ? signature.substring(0, 3)
+      : signature.substring(0, 5);
+    if (Object.prototype.hasOwnProperty.call(prefix, pref))
+      return Buffer.from(
+        b58cdecode(
+          signature,
+          Object.getOwnPropertyDescriptor(prefix, pref)?.value,
+        ),
+      ).toString("hex");
+    throw new Error("Unsupported Signature Type");
+  }
+
+  async estimateInitiateFees(source: string): Promise<number> {
+    const dummyTx = {
+      receivingAddress: "tz1aKTCbAUuea2RV9kxqRVRg3HT7f1RKnp6a",
+      secretHash:
+        "169cbd29345af89a0983f28254e71bdd1367890b9876fc8a9ea117c32f6a521b",
+      refundTimestamp: 1702043022,
+      rewardForRedeem: 0,
+      netAmount: 100,
+    };
+
+    const tx = this.buildInitiateTransaction(dummyTx);
+    if (tx.amount == undefined) tx.amount = 0;
+
+    const header = await this._tezos.rpc.getBlockHeader();
+    const contract = await this._tezos.rpc.getContract(source);
+    const op = await this._tezos.rpc.runOperation({
+      chain_id: header.chain_id,
+      operation: {
+        branch: header.hash,
+        signature:
+          "sigUHx32f9wesZ1n2BWpixXz4AQaZggEtchaQNHYGRCoWNAXx45WGW2ua3apUUUAGMLPwAU41QoaFCzVSL61VaessLg4YbbP",
+        contents: [
+          {
+            amount: tx.amount.toString(),
+            counter: (parseInt(contract.counter || "0") + 1).toString(),
+            destination: this._contractAddress,
+            fee: "10000",
+            gas_limit: this._gasLimit.toString(),
+            kind: OpKind.TRANSACTION,
+            source: "tz1Y8UNsMSCXyDgma8Ya51eLx8Qu4AoLm8vt",
+            storage_limit: "300",
+            parameters: tx.data,
+          },
+        ],
+      },
+    });
+
+    let paidStorageDiff = 0,
+      consumedGas = 0;
+
+    (op.contents as OperationContentsAndResultTransaction[]).forEach((tx) => {
+      consumedGas += parseInt(tx.metadata.operation_result.consumed_gas || "0");
+      paidStorageDiff += parseInt(
+        tx.metadata.operation_result.paid_storage_size_diff || "0",
+      );
+    });
+
+    return (
+      this._minimalFees +
+      this._minimalNanotezPerGasUnit * consumedGas +
+      this._minimalNanotezPerByte * this._initiateTxSize +
+      paidStorageDiff * 1000 * this._minimalNanotezPerByte
+    );
+  }
+
+  async estimateRedeemFees(): Promise<RedeemFees> {
```

Comment:
accept receiving address
Check if address balance is non zero, otherwise add 257 * costPerByte to the fee